### PR TITLE
irept: use single map for all named sub-nodes

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — DISMISSED** [review]
This is probably a very good idea, with two requests:
1) The version number of goto binaries needs to be incremented.
2) We need to do benchmarking, most likely this is a major improvement (it's just the `operator==` that's of concern).

**Comment by smowton** [comment]
The hash-container change suggests we'll always include comments in a particular irep's hashcode? In that case surely unordered_map will break as hash-code equality and operator== will disagree?

**Comment by kroening** [comment]
@smowton Was still getting there; they now agree.

**Comment by kroening** [comment]
One more BTW @tautschnig : The goto-binary format is tolerant of this change (forwards and backwards), i.e., no need to tweak the version number for this.

File: src/util/irep_serialization.cpp [hunk 1]
```diff
@@ -38,13 +38,6 @@ void irep_serializationt::write_irep(
     reference_convert(it->second, out);
   }
 
-  forall_named_irep(it, irep.get_comments())
-  {
-    out.put('C');
```
Comment (tautschnig) [inline src/util/irep_serialization.cpp:43]: Since we no longer put this, reading a binary generated by a newer version of `goto-cc` using an older version will place all comments in the non-comment map. Thus I think a change to the goto binary version is required, but maybe I'm getting this wrong.

Comment (kroening) [reply src/util/irep_serialization.cpp:43]: No, the reading in old versions happens using .add, which will put everything into the right map.

**Comment by tautschnig** [comment]
I will take care of benchmarking once CI passes.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ced9234e53281b7f76666325c26aaf0983f8359d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92991087

**Review by tautschnig — CHANGES_REQUESTED** [review]
With this change there is a desirable reduction in memory consumption, but also approximately 20% slowdown across all benchmarks. This additional time is spent in `irept::number_of_non_comments`. The two dominating causes of invoking this are 1) computing a hash - to be optimised away via `HASH_CODE`, and 2) sorting operands in expression simplification - see #1997 for removing it.

Overall I think we should work to enable this optimisation, but we have more work to do to avoid a performance regression.

**Comment by kroening** [comment]
@tautschnig Many thanks -- I've just done a tweak that should reduce the cost of counting for the case of hashing. May I ask you to do another run?

**Comment by tautschnig** [comment]
> May I ask you to do another run?

Running.

**Comment by tautschnig** [comment]
screen shot 2018-12-05 at 13 55 56

That's the quantile plot for ReachSafety-BitVectors - it seems the new version is back en par with develop.

File: src/goto-programs/show_goto_functions_json.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ json_objectt show_goto_functions_jsont::convert(
   const goto_functionst &goto_functions)
 {
   json_arrayt json_functions;
-  const json_irept no_comments_irep_converter(false);
+  const json_irept no_comments_irep_converter;
```
Comment (tautschnig) [inline src/goto-programs/show_goto_functions_json.cpp]: Does this still do what the name promises? If so, how?

File: src/util/irep.cpp [hunk 3]
```diff
@@ -502,8 +510,8 @@ int irept::compare(const irept &i) const
   if(r!=0)
     return r;
 
-  const subt::size_type size=get_sub().size(),
-        i_size=i.get_sub().size();
+  const auto size = get_sub().size(), i_size = i.get_sub().size();
+
```
Comment (tautschnig) [inline src/util/irep.cpp:515]: I believe this is an unrelated change. Let's try to keep changes to an absolute minimum.

File: src/util/irep.cpp [hunk 4]
```diff
@@ -537,24 +546,45 @@ int irept::compare(const irept &i) const
   {
     irept::named_subt::const_iterator it1, it2;
 
-    for(it1=get_named_sub().begin(),
-        it2=i.get_named_sub().begin();
-        it1!=get_named_sub().end() && it2!=i.get_named_sub().end();
-        it1++,
-        it2++)
+    for(it1 = get_named_sub().begin(),
+    it2 = i.get_named_sub().begin();
```
Comment (tautschnig) [inline src/util/irep.cpp]: Indent looks strange.

File: src/util/irep.cpp [hunk 5]
```diff
@@ -537,24 +546,45 @@ int irept::compare(const irept &i) const
...
+        it2++;
+        continue;
+      }
+
+      // the case that both it1 and it2 are .end() is treated
+      // by the loop guard; furthermore, the number of non-comments
+      // must be the same
+      INVARIANT(it1 != get_named_sub().end(), "not at end of get_named_sub()");
+      INVARIANT(
+        it2 != i.get_named_sub().end(), "not at end of i.get_named_sub()");
```
Comment (tautschnig) [inline src/util/irep.cpp:575]: This should basically be the same code as `irept::operator==` but does extra checks and just overall looks a bit different. Maybe we can avoid this?

File: src/util/json_irep.cpp [hunk 6]
```diff
@@ -38,10 +35,8 @@ json_objectt json_irept::convert_from_irep(const irept &irep) const
     irep_object["id"]=json_stringt(irep.id_string());
 
   convert_sub_tree("sub", irep.get_sub(), irep_object);
-  convert_named_sub_tree("namedSub", irep.get_named_sub(), irep_object);
 
-  if(include_comments)
-    convert_named_sub_tree("comment", irep.get_comments(), irep_object);
```
Comment (tautschnig) [inline src/util/json_irep.cpp:44]: That's not longer present. May or may not matter to any code consuming the JSON output.

File: src/util/json_irep.cpp [hunk 7]
```diff
@@ -19,10 +19,7 @@ Author: Thomas Kiley, thomas.kiley@diffblue.com
 
 /// To convert to JSON from an irep structure by recursively generating JSON
 /// for the different sub trees.
```
Comment (tautschnig) [inline src/util/json_irep.cpp]: This comment should actually go with the class, not with the constructor.

File: src/util/json_irep.h [hunk 8]
```diff
@@ -20,7 +20,7 @@ class json_objectt;
 class json_irept
 {
 public:
-  explicit json_irept(bool include_comments);
+  json_irept();
```
Comment (tautschnig) [inline src/util/json_irep.h]: Could now simply be removed.

File: src/util/xml_irep.cpp [hunk 9]
```diff
@@ -35,13 +35,6 @@ void convert(
     x_nsub.set_attribute("name", name2string(it->first));
     convert(it->second, x_nsub);
   }
-
-  forall_named_irep(it, irep.get_comments())
-  {
-    xmlt &x_com = xml.new_element("comment");
```
Comment (tautschnig) [inline src/util/xml_irep.cpp:41]: No longer available in the output.

File: unit/util/irep.cpp [hunk 10]
```diff
@@ -145,7 +143,7 @@ SCENARIO("irept_memory", "[core][utils][irept]")
 
       irep.add("#a_comment", irep2);
       REQUIRE(irep.find("#a_comment").id() == "second_irep");
-      REQUIRE(irep.get_comments().size() == 1);
+      REQUIRE(irep.get_named_sub().size() == 2);
```
Comment (tautschnig) [inline unit/util/irep.cpp:148]: We should now include `is_comment` in this unit test (the test aims to cover the full API).

File: unit/util/irep.cpp [hunk 11]
```diff
@@ -159,19 +157,17 @@ SCENARIO("irept_memory", "[core][utils][irept]")
 
       irep.move_to_named_sub("another_entry", irep2);
       REQUIRE(irep.get_sub().size() == 1);
-      REQUIRE(irep.get_named_sub().size() == 1);
-      REQUIRE(irep.get_comments().size() == 1);
+      REQUIRE(irep.get_named_sub().size() == 2);
```
Comment (tautschnig) [inline unit/util/irep.cpp:163]: Could at least call `number_of_non_comments`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d202f7c3b8b2dfed1cdd7a95c537ac36e26ad45d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93671584

File: src/goto-programs/show_goto_functions_json.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-programs/show_goto_functions_json.cpp]: removed these changes

File: src/util/irep.cpp (hunk 3 above)
Comment (kroening) [reply src/util/irep.cpp:515]: ok

File: src/util/irep.cpp (hunk 4 above)
Comment (kroening) [reply src/util/irep.cpp]: clang-format disabled

File: src/util/irep.cpp (hunk 5 above)
Comment (kroening) [reply src/util/irep.cpp:575]: could use compare() for both, but seems to be an unrelated change

File: src/util/json_irep.cpp (hunk 6 above)
Comment (kroening) [reply src/util/json_irep.cpp:44]: removed

File: src/util/xml_irep.cpp (hunk 9 above)
Comment (kroening) [reply src/util/xml_irep.cpp:41]: restored

File: unit/util/irep.cpp (hunk 10 above)
Comment (kroening) [reply unit/util/irep.cpp:148]: ok

File: unit/util/irep.cpp (hunk 11 above)
Comment (kroening) [reply unit/util/irep.cpp:163]: done

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 602486189a8bf43a4fc97d16c94d6b9567d9d252).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93968598
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: df4dee099b7449fb3cc91f2d516316dd2956a6b9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93969059
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 31ab10ae6ad74e3a3900c50038a08d3e3bf3cb9b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93970956

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6933585745347eb6577193d577f34ae1a485adf8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95243811

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 48c40db564b1af55b3f289122872d34efdafcbe8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95336240

**Comment by tautschnig** [comment]
I am [attaching](https://github.com/diffblue/cbmc/files/2698253/results.2018-12-20_0755.table.html.zip) the results for develop vs this branch on SV-COMP's ReachSafety-ECA category, which is the category where we spend the most time actually running symbolic execution (before handing the formula to the SAT solver). The difference is very small, but overall in favour of the proposed change. Some tasks are slower, some finish quicker. The total memory consumption is lower, as expected.

**Review by tautschnig — APPROVED** [review]
